### PR TITLE
Add REST for tso microservice

### PR DESCRIPTION
**Pull request by hnes**
### What problem does this PR solve?

Issue Number: ref #5836

### What is changed and how does it work?

```commit-message
```

Add REST into tso microservice

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+// FIXME: details of input json body params
+// @Tags     admin
+// @Summary  Reset the ts.
+// @Accept   json
+// @Param    body  body  object  true  "json params"
+// @Produce  json
+// @Success  200  {string}  string  "Reset ts successfully."
+// @Failure  400  {string}  string  "The input is invalid."
+// @Failure  403  {string}  string  "Reset ts is forbidden."
+// @Failure  500  {string}  string  "PD server failed to proceed the request."
```
Comment (binshi-bing) [inline pkg/mcs/tso/server/apis/v1/api.go:103]: **TSO** server failed to proceed the request

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+			s.rd.JSON(w, http.StatusBadRequest, "invalid force-use-larger value")
+			return
+		}
+	}
+	var ignoreSmaller, skipUpperBoundCheck bool
+	if forceUseLarger {
+		ignoreSmaller, skipUpperBoundCheck = true, true
+	}
+
+	if err = handler.ResetTS(ts, ignoreSmaller, skipUpperBoundCheck); err != nil {
```
Comment (binshi-bing) [inline pkg/mcs/tso/server/apis/v1/api.go:144]: We should add support for keyspace in ResetTS. Where do we pass keyspaceID and handle keyspaceID?

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+// if force-use-larger=true:
+//
+//	reset ts to max(current ts, input ts).
+//
+// else:
+//
+//	reset ts to input ts if it > current ts and < upper bound, error if not in that range
+//
+// during EBS based restore, we call this to make sure ts of pd >= resolved_ts in backup.
+func (s *Service) ResetTS(w http.ResponseWriter, r *http.Request) {
```
Comment (binshi-bing) [inline pkg/mcs/tso/server/apis/v1/api.go:114]: It seems that there is no need to duplicate ResetTS() in server/api/admin.go. Some common admin RESTful APIs can be put into pkg/tso/api/tsoadmin.go.

For tso service, we execute the following code somewhere (mightn't be in router.go)
tsoAdminHandler := newTSOAdminHandler(svr, rd)
registerFunc(apiRouter, "/admin/reset-ts", tsoAdminHandler.ResetTS, setMethods(http.MethodPost), setAuditBackend(localLog, prometheus))

**Review by binshi-bing — COMMENTED** [review]
left some comments

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		IndentJSON: true,
+	})
+}
+
+// NewService returns a new Service.
+func NewService(srv *tsoserver.Service) *Service {
+	apiHandlerEngine := gin.New()
+	apiHandlerEngine.Use(gin.Recovery())
+	apiHandlerEngine.Use(cors.Default())
+	apiHandlerEngine.Use(gzip.Gzip(gzip.DefaultCompression))
```
Comment (nolouch) [inline pkg/mcs/tso/server/apis/v1/api.go:69]: Maybe need the `ServiceRedirector` middleware like https://github.com/tikv/pd/issues/6051.

File: pkg/mcs/tso/server/apis/v1/api.go (hunk 3 above)
Comment (binshi-bing) [reply pkg/mcs/tso/server/apis/v1/api.go:114]: If it needs more time, let's ignore it for now and check-in this pr first.

File: pkg/mcs/tso/server/apis/v1/api.go (hunk 2 above)
Comment (binshi-bing) [reply pkg/mcs/tso/server/apis/v1/api.go:144]: Let's ignore keyspace for now.

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+)
+
+func init() {
+	tsoserver.SetUpRestHandler = func(srv *tsoserver.Service) (http.Handler, apiutil.APIServiceGroup) {
+		s := NewService(srv)
+		return s.handler(), apiServiceGroup
+	}
+}
+
+// Service is the resource group service.
```
[the remark this answers is not in the document]
Comment (hnes) [reply pkg/mcs/tso/server/apis/v1/api.go:49]: Thanks a lot.

File: pkg/mcs/tso/server/apis/v1/api.go (hunk 2 above)
Comment (hnes) [reply pkg/mcs/tso/server/apis/v1/api.go:144]: OK.

File: pkg/mcs/tso/server/apis/v1/api.go (hunk 4 above)
Comment (hnes) [reply pkg/mcs/tso/server/apis/v1/api.go:69]: OK. Thanks a lot :-)

File: pkg/mcs/tso/server/apis/v1/api.go (hunk 3 above)
Comment (hnes) [reply pkg/mcs/tso/server/apis/v1/api.go]: Very good idea 👍

Comment (hnes) [reply pkg/mcs/tso/server/apis/v1/api.go]: Please review the latest implementation.

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 TiKV Project Authors.
```
Comment (rleungx) [inline pkg/mcs/tso/server/apis/v1/api.go]:
```suggestion
// Copyright 2023 TiKV Project Authors.
```

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/gin-gonic/gin"
+	tsoserver "github.com/tikv/pd/pkg/mcs/tso/server"
+	"github.com/tikv/pd/pkg/tso"
+	"github.com/tikv/pd/pkg/utils/apiutil"
+	"github.com/tikv/pd/pkg/utils/apiutil/multiservicesapi"
+	"github.com/unrolled/render"
+)
+
+// APIPathPrefix is the prefix of the API path.
+const APIPathPrefix = "/"
```
Comment (rleungx) [inline pkg/mcs/tso/server/apis/v1/api.go:31]: Should it be `/tso/api/v1/`?

File: cmd/pd-server/main.go [hunk 8]
```diff
@@ -28,6 +28,7 @@ import (
 	"github.com/tikv/pd/pkg/errs"
 	resource_manager "github.com/tikv/pd/pkg/mcs/resource_manager/server"
 	tso "github.com/tikv/pd/pkg/mcs/tso/server"
+	_ "github.com/tikv/pd/pkg/mcs/tso/server/apis/v1" // init tso API group
```
Comment (rleungx) [inline cmd/pd-server/main.go:31]: move it to server?

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// NewService returns a new Service.
+func NewService(srv *tsoserver.Service) *Service {
+	apiHandlerEngine := gin.New()
+	apiHandlerEngine.Use(gin.Recovery())
+	apiHandlerEngine.Use(cors.Default())
+	apiHandlerEngine.Use(gzip.Gzip(gzip.DefaultCompression))
+	apiHandlerEngine.Use(func(c *gin.Context) {
+		c.Set("service", srv.GetBasicServer())
```
Comment (rleungx) [inline pkg/mcs/tso/server/apis/v1/api.go:72]:
```suggestion
		c.Set("service", srv)
```

**Comment by rleungx** [comment]
How about adding some tests in `tests/mcs`?

File: cmd/pd-server/main.go (hunk 8 above)
Comment (hnes) [reply cmd/pd-server/main.go:31]: OK.

File: pkg/mcs/tso/server/apis/v1/api.go (hunk 7 above)
Comment (hnes) [reply pkg/mcs/tso/server/apis/v1/api.go:31]: Thanks a lot.

File: pkg/mcs/tso/server/apis/v1/api.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/tikv/pd/pkg/utils/apiutil"
+	"github.com/tikv/pd/pkg/utils/apiutil/multiservicesapi"
+	"github.com/unrolled/render"
+)
+
+// APIPathPrefix is the prefix of the API path.
+const APIPathPrefix = "/tso/api/v1"
+
+var (
+	apiServiceGroup = apiutil.APIServiceGroup{
```
Comment (nolouch) [inline pkg/mcs/tso/server/apis/v1/api.go:34]: the tools need switch to use new API path? how to make the tools can works within the multi-services mode?

**Review by binshi-bing — APPROVED** [review]
Great Job!

File: pkg/mcs/tso/server/apis/v1/api.go (hunk 10 above)
Comment (hnes) [reply pkg/mcs/tso/server/apis/v1/api.go:34]: Very good point. Thanks a lot. I would solve this problem in the next PR.

**Review by rleungx — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**